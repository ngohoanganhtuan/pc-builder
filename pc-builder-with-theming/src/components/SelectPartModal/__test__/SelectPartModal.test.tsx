import { render, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';

// Hooks
import { useGetData } from '@hooks';

// Components
import { SelectPartModal } from '@components';

// Enums
import { BRAND_NAMES, PC_PART_TYPES } from '@enums';

const mockData = Array.from({ length: 12 }).map((mockItem, index) => ({
  thumbnail:
    'https://cdn.techzones.vn/Data/Sites/1/Product/35710/techzones-amd-ryzen-7-5600x-6c12t-32mb-cache-37ghz-up-to-46ghz.jpg',
  name: `${index} AMD Ryzen 7 5600X - 6C/12T 32MB Cache 3.7GHz Up to 4.6GHz`,
  price: 7600000,
  type: PC_PART_TYPES.CPU,
  brandName: BRAND_NAMES.AMD,
  id: index,
}));

jest.mock('@hooks', () => {
  return {
    useGetData: jest.fn(),
  };
});

describe('Test [SelectPartModal] component', () => {
  const handleCloseModal = jest.fn();
  const handleSelectPCPart = jest.fn();
  const initialProps = {
    handleCloseModal,
    handleSelectPCPart,
  };

  afterEach(() => {
    handleCloseModal.mockClear();
    handleSelectPCPart.mockClear();
  });

  test('Component [SelectPartModal] should render correctly when data is fetching', () => {
    const mockHook = (useGetData as jest.Mock).mockImplementation(() => ({
      data: mockData,
      isFetching: true,
    }));
    const component = renderer
      .create(
        <BrowserRouter>
          <SelectPartModal {...initialProps} />
        </BrowserRouter>
      )
      .toJSON();
    expect(component).toMatchSnapshot();
    mockHook.mockReset();
  });

  test('Component [SelectPartModal] should render correctly when data is fetched', () => {
    (useGetData as jest.Mock).mockImplementation(() => ({
      data: mockData,
      isFetching: false,
    }));
    const component = renderer
      .create(
        <BrowserRouter>
          <SelectPartModal {...initialProps} />
        </BrowserRouter>
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  test('Component [SelectPartModal] calls function close modal when user click close button', () => {
    const { getByText } = render(
      <BrowserRouter>
        <SelectPartModal {...initialProps} />
      </BrowserRouter>
    );
    const button = getByText('Close');
    expect(button.textContent).toEqual('Close');
    fireEvent.click(button);
    expect(handleCloseModal).toHaveBeenCalled();
  });

  test('Component [SelectPartModal] calls function select pc part when user click select button', () => {
    const { getAllByTestId } = render(
      <BrowserRouter>
        <SelectPartModal {...initialProps} />
      </BrowserRouter>
    );
    const button = getAllByTestId('pc-part-select-button')[0];
    fireEvent.click(button);
    expect(handleSelectPCPart).toHaveBeenCalled();
  });

  test('Component [SelectPartModal] should show load more button if hook returns more than 10 items', () => {
    const { getByTestId } = render(
      <BrowserRouter>
        <SelectPartModal {...initialProps} />
      </BrowserRouter>
    );
    expect(getByTestId('load-more-button')).toBeInTheDocument();
  });
});
