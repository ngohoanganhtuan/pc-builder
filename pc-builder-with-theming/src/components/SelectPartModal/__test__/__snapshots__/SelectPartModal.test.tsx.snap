// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test [SelectPartModal] component Component [SelectPartModal] should render correctly when data is fetched 1`] = `
<div
  className="modal d-block select-part-modal"
>
  <div
    className="modal-dialog modal-xl modal-dialog-scrollable"
  >
    <div
      className="modal-content shadow"
    >
      <div
        className="modal-body row overflow-hidden"
      >
        <div
          className="col-12 col-lg-3 d-flex flex-column pb-3 justify-content-between"
        >
          <div
            className="d-flex flex-column filter"
          >
            <div
              className="mb-3"
            >
              <p
                className="form-label filter-brand"
              >
                Brand
              </p>
              <div
                className="d-flex flex-column flex-wrap"
              >
                <div
                  className="form-check"
                >
                  <input
                    checked={false}
                    className="form-check-input"
                    id="brand-checkbox-0"
                    name="brand-checkbox-0"
                    onChange={[Function]}
                    type="checkbox"
                    value="AMD"
                  />
                  <label
                    className="form-check-label brand-checkbox-label"
                    htmlFor="brand-checkbox-0"
                  >
                    AMD
                  </label>
                </div>
              </div>
            </div>
            <div
              className="mb-3"
            >
              <p
                className="form-label filter-price"
              >
                Price range
              </p>
              <div
                className="range-slider d-flex flex-row justify-content-between align-items-center gap-2"
              >
                <input
                  className="form-control"
                  data-testid="min-range-input"
                  onChange={[Function]}
                  placeholder="Minimum"
                  value={0}
                />
                ~
                <input
                  className="form-control"
                  data-testid="max-range-input"
                  onChange={[Function]}
                  placeholder="Maximum"
                  value={0}
                />
              </div>
            </div>
          </div>
          <button
            className="btn btn-danger"
            data-testid="close-modal-button"
            onClick={[Function]}
          >
            Close
          </button>
        </div>
        <div
          className="col-12 col-lg-9 d-flex flex-column gap-3 pc-parts-list pb-3"
        >
          <div
            className="shadow d-grid pc-part"
          >
            <div
              className="p-1"
            >
              <img
                className="border rounded-circle pc-part-thumbnail"
                data-testid="pc-part-thumbnail"
                height="100%"
                src="https://cdn.techzones.vn/Data/Sites/1/Product/35710/techzones-amd-ryzen-7-5600x-6c12t-32mb-cache-37ghz-up-to-46ghz.jpg"
                width="100%"
              />
            </div>
            <div
              className="p-3 d-flex flex-column align-items-start justify-content-between"
            >
              <a
                className="name card-title fs-6 text-dark fw-bold"
                href="/detail/0"
                onClick={[Function]}
              >
                0 AMD Ryzen 7 5600X - 6C/12T 32MB Cache 3.7GHz Up to 4.6GHz
              </a>
            </div>
            <div
              className="p-3 d-flex flex-column align-items-end justify-content-between"
            >
              <p
                className="fs-6 price"
              >
                7.600.000 ₫
              </p>
              <button
                className="btn btn-sm btn-success shadow"
                data-testid="pc-part-select-button"
                onClick={[Function]}
              >
                Select
              </button>
            </div>
          </div>
          <div
            className="shadow d-grid pc-part"
          >
            <div
              className="p-1"
            >
              <img
                className="border rounded-circle pc-part-thumbnail"
                data-testid="pc-part-thumbnail"
                height="100%"
                src="https://cdn.techzones.vn/Data/Sites/1/Product/35710/techzones-amd-ryzen-7-5600x-6c12t-32mb-cache-37ghz-up-to-46ghz.jpg"
                width="100%"
              />
            </div>
            <div
              className="p-3 d-flex flex-column align-items-start justify-content-between"
            >
              <a
                className="name card-title fs-6 text-dark fw-bold"
                href="/detail/1"
                onClick={[Function]}
              >
                1 AMD Ryzen 7 5600X - 6C/12T 32MB Cache 3.7GHz Up to 4.6GHz
              </a>
            </div>
            <div
              className="p-3 d-flex flex-column align-items-end justify-content-between"
            >
              <p
                className="fs-6 price"
              >
                7.600.000 ₫
              </p>
              <button
                className="btn btn-sm btn-success shadow"
                data-testid="pc-part-select-button"
                onClick={[Function]}
              >
                Select
              </button>
            </div>
          </div>
          <div
            className="shadow d-grid pc-part"
          >
            <div
              className="p-1"
            >
              <img
                className="border rounded-circle pc-part-thumbnail"
                data-testid="pc-part-thumbnail"
                height="100%"
                src="https://cdn.techzones.vn/Data/Sites/1/Product/35710/techzones-amd-ryzen-7-5600x-6c12t-32mb-cache-37ghz-up-to-46ghz.jpg"
                width="100%"
              />
            </div>
            <div
              className="p-3 d-flex flex-column align-items-start justify-content-between"
            >
              <a
                className="name card-title fs-6 text-dark fw-bold"
                href="/detail/2"
                onClick={[Function]}
              >
                2 AMD Ryzen 7 5600X - 6C/12T 32MB Cache 3.7GHz Up to 4.6GHz
              </a>
            </div>
            <div
              className="p-3 d-flex flex-column align-items-end justify-content-between"
            >
              <p
                className="fs-6 price"
              >
                7.600.000 ₫
              </p>
              <button
                className="btn btn-sm btn-success shadow"
                data-testid="pc-part-select-button"
                onClick={[Function]}
              >
                Select
              </button>
            </div>
          </div>
          <div
            className="shadow d-grid pc-part"
          >
            <div
              className="p-1"
            >
              <img
                className="border rounded-circle pc-part-thumbnail"
                data-testid="pc-part-thumbnail"
                height="100%"
                src="https://cdn.techzones.vn/Data/Sites/1/Product/35710/techzones-amd-ryzen-7-5600x-6c12t-32mb-cache-37ghz-up-to-46ghz.jpg"
                width="100%"
              />
            </div>
            <div
              className="p-3 d-flex flex-column align-items-start justify-content-between"
            >
              <a
                className="name card-title fs-6 text-dark fw-bold"
                href="/detail/3"
                onClick={[Function]}
              >
                3 AMD Ryzen 7 5600X - 6C/12T 32MB Cache 3.7GHz Up to 4.6GHz
              </a>
            </div>
            <div
              className="p-3 d-flex flex-column align-items-end justify-content-between"
            >
              <p
                className="fs-6 price"
              >
                7.600.000 ₫
              </p>
              <button
                className="btn btn-sm btn-success shadow"
                data-testid="pc-part-select-button"
                onClick={[Function]}
              >
                Select
              </button>
            </div>
          </div>
          <div
            className="shadow d-grid pc-part"
          >
            <div
              className="p-1"
            >
              <img
                className="border rounded-circle pc-part-thumbnail"
                data-testid="pc-part-thumbnail"
                height="100%"
                src="https://cdn.techzones.vn/Data/Sites/1/Product/35710/techzones-amd-ryzen-7-5600x-6c12t-32mb-cache-37ghz-up-to-46ghz.jpg"
                width="100%"
              />
            </div>
            <div
              className="p-3 d-flex flex-column align-items-start justify-content-between"
            >
              <a
                className="name card-title fs-6 text-dark fw-bold"
                href="/detail/4"
                onClick={[Function]}
              >
                4 AMD Ryzen 7 5600X - 6C/12T 32MB Cache 3.7GHz Up to 4.6GHz
              </a>
            </div>
            <div
              className="p-3 d-flex flex-column align-items-end justify-content-between"
            >
              <p
                className="fs-6 price"
              >
                7.600.000 ₫
              </p>
              <button
                className="btn btn-sm btn-success shadow"
                data-testid="pc-part-select-button"
                onClick={[Function]}
              >
                Select
              </button>
            </div>
          </div>
          <div
            className="shadow d-grid pc-part"
          >
            <div
              className="p-1"
            >
              <img
                className="border rounded-circle pc-part-thumbnail"
                data-testid="pc-part-thumbnail"
                height="100%"
                src="https://cdn.techzones.vn/Data/Sites/1/Product/35710/techzones-amd-ryzen-7-5600x-6c12t-32mb-cache-37ghz-up-to-46ghz.jpg"
                width="100%"
              />
            </div>
            <div
              className="p-3 d-flex flex-column align-items-start justify-content-between"
            >
              <a
                className="name card-title fs-6 text-dark fw-bold"
                href="/detail/5"
                onClick={[Function]}
              >
                5 AMD Ryzen 7 5600X - 6C/12T 32MB Cache 3.7GHz Up to 4.6GHz
              </a>
            </div>
            <div
              className="p-3 d-flex flex-column align-items-end justify-content-between"
            >
              <p
                className="fs-6 price"
              >
                7.600.000 ₫
              </p>
              <button
                className="btn btn-sm btn-success shadow"
                data-testid="pc-part-select-button"
                onClick={[Function]}
              >
                Select
              </button>
            </div>
          </div>
          <div
            className="shadow d-grid pc-part"
          >
            <div
              className="p-1"
            >
              <img
                className="border rounded-circle pc-part-thumbnail"
                data-testid="pc-part-thumbnail"
                height="100%"
                src="https://cdn.techzones.vn/Data/Sites/1/Product/35710/techzones-amd-ryzen-7-5600x-6c12t-32mb-cache-37ghz-up-to-46ghz.jpg"
                width="100%"
              />
            </div>
            <div
              className="p-3 d-flex flex-column align-items-start justify-content-between"
            >
              <a
                className="name card-title fs-6 text-dark fw-bold"
                href="/detail/6"
                onClick={[Function]}
              >
                6 AMD Ryzen 7 5600X - 6C/12T 32MB Cache 3.7GHz Up to 4.6GHz
              </a>
            </div>
            <div
              className="p-3 d-flex flex-column align-items-end justify-content-between"
            >
              <p
                className="fs-6 price"
              >
                7.600.000 ₫
              </p>
              <button
                className="btn btn-sm btn-success shadow"
                data-testid="pc-part-select-button"
                onClick={[Function]}
              >
                Select
              </button>
            </div>
          </div>
          <div
            className="shadow d-grid pc-part"
          >
            <div
              className="p-1"
            >
              <img
                className="border rounded-circle pc-part-thumbnail"
                data-testid="pc-part-thumbnail"
                height="100%"
                src="https://cdn.techzones.vn/Data/Sites/1/Product/35710/techzones-amd-ryzen-7-5600x-6c12t-32mb-cache-37ghz-up-to-46ghz.jpg"
                width="100%"
              />
            </div>
            <div
              className="p-3 d-flex flex-column align-items-start justify-content-between"
            >
              <a
                className="name card-title fs-6 text-dark fw-bold"
                href="/detail/7"
                onClick={[Function]}
              >
                7 AMD Ryzen 7 5600X - 6C/12T 32MB Cache 3.7GHz Up to 4.6GHz
              </a>
            </div>
            <div
              className="p-3 d-flex flex-column align-items-end justify-content-between"
            >
              <p
                className="fs-6 price"
              >
                7.600.000 ₫
              </p>
              <button
                className="btn btn-sm btn-success shadow"
                data-testid="pc-part-select-button"
                onClick={[Function]}
              >
                Select
              </button>
            </div>
          </div>
          <div
            className="shadow d-grid pc-part"
          >
            <div
              className="p-1"
            >
              <img
                className="border rounded-circle pc-part-thumbnail"
                data-testid="pc-part-thumbnail"
                height="100%"
                src="https://cdn.techzones.vn/Data/Sites/1/Product/35710/techzones-amd-ryzen-7-5600x-6c12t-32mb-cache-37ghz-up-to-46ghz.jpg"
                width="100%"
              />
            </div>
            <div
              className="p-3 d-flex flex-column align-items-start justify-content-between"
            >
              <a
                className="name card-title fs-6 text-dark fw-bold"
                href="/detail/8"
                onClick={[Function]}
              >
                8 AMD Ryzen 7 5600X - 6C/12T 32MB Cache 3.7GHz Up to 4.6GHz
              </a>
            </div>
            <div
              className="p-3 d-flex flex-column align-items-end justify-content-between"
            >
              <p
                className="fs-6 price"
              >
                7.600.000 ₫
              </p>
              <button
                className="btn btn-sm btn-success shadow"
                data-testid="pc-part-select-button"
                onClick={[Function]}
              >
                Select
              </button>
            </div>
          </div>
          <div
            className="shadow d-grid pc-part"
          >
            <div
              className="p-1"
            >
              <img
                className="border rounded-circle pc-part-thumbnail"
                data-testid="pc-part-thumbnail"
                height="100%"
                src="https://cdn.techzones.vn/Data/Sites/1/Product/35710/techzones-amd-ryzen-7-5600x-6c12t-32mb-cache-37ghz-up-to-46ghz.jpg"
                width="100%"
              />
            </div>
            <div
              className="p-3 d-flex flex-column align-items-start justify-content-between"
            >
              <a
                className="name card-title fs-6 text-dark fw-bold"
                href="/detail/9"
                onClick={[Function]}
              >
                9 AMD Ryzen 7 5600X - 6C/12T 32MB Cache 3.7GHz Up to 4.6GHz
              </a>
            </div>
            <div
              className="p-3 d-flex flex-column align-items-end justify-content-between"
            >
              <p
                className="fs-6 price"
              >
                7.600.000 ₫
              </p>
              <button
                className="btn btn-sm btn-success shadow"
                data-testid="pc-part-select-button"
                onClick={[Function]}
              >
                Select
              </button>
            </div>
          </div>
          <button
            className="btn btn-danger"
            data-testid="load-more-button"
            onClick={[Function]}
          >
            Load more
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Test [SelectPartModal] component Component [SelectPartModal] should render correctly when data is fetching 1`] = `
<div
  className="modal d-block select-part-modal"
>
  <div
    className="modal-dialog modal-xl modal-dialog-scrollable"
  >
    <div
      className="modal-content shadow"
    >
      <div
        className="modal-body row overflow-hidden"
      >
        <div
          className="col-12 col-lg-3 d-flex flex-column pb-3 justify-content-between"
        >
          <div
            className="d-flex flex-column filter"
          >
            <div
              className="mb-3"
            >
              <p
                className="form-label filter-brand"
              >
                Brand
              </p>
              <div
                className="d-flex flex-column flex-wrap"
              >
                <div
                  className="form-check"
                >
                  <input
                    checked={false}
                    className="form-check-input"
                    id="brand-checkbox-0"
                    name="brand-checkbox-0"
                    onChange={[Function]}
                    type="checkbox"
                    value="AMD"
                  />
                  <label
                    className="form-check-label brand-checkbox-label"
                    htmlFor="brand-checkbox-0"
                  >
                    AMD
                  </label>
                </div>
              </div>
            </div>
            <div
              className="mb-3"
            >
              <p
                className="form-label filter-price"
              >
                Price range
              </p>
              <div
                className="range-slider d-flex flex-row justify-content-between align-items-center gap-2"
              >
                <input
                  className="form-control"
                  data-testid="min-range-input"
                  onChange={[Function]}
                  placeholder="Minimum"
                  value={0}
                />
                ~
                <input
                  className="form-control"
                  data-testid="max-range-input"
                  onChange={[Function]}
                  placeholder="Maximum"
                  value={0}
                />
              </div>
            </div>
          </div>
          <button
            className="btn btn-danger"
            data-testid="close-modal-button"
            onClick={[Function]}
          >
            Close
          </button>
        </div>
        <div
          className="col-12 col-lg-9 d-flex flex-column gap-3 pc-parts-list pb-3"
        >
          <div
            className="spinner-border align-self-center m-auto"
            role="status"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
